app/models/concerns/reviewable_action_builder.rb

```diff
@@ -3,81 +3,71 @@
 module ReviewableActionBuilder
   extend ActiveSupport::Concern
 
+  attr_accessor :actions, :guardian, :action_args
+
   # Standard post-actions bundle. Consumers should use the returned
   # bundle value when adding post-focused actions.
   #
-  # @param actions [Reviewable::Actions] Actions instance to add the bundle to.
-  # @param guardian [Guardian] Guardian instance to check permissions.
-  #
   # @return [Reviewable::Actions::Bundle] The created post actions bundle.
-  def build_post_actions_bundle(actions, guardian)
-    bundle =
-      actions.add_bundle(
-        "#{id}-post-actions",
-        label: "reviewables.actions.post_actions.bundle_title",
-      )
-
-    # Always include the no-op action
-    build_action(actions, :no_action_post, bundle:)
+  def build_post_actions_bundle
+    bundle_actions = { no_action_post: {} }
+    if target_post
+      if target_post.trashed? && @guardian.can_recover_post?(target_post)
+        bundle_actions[:restore_post] = {}
+      end
 
-    return bundle unless target_post
+      if target_post.hidden?
+        bundle_actions[:unhide_post] = {} if !target_post.user_deleted?
```

Comment:
Shall we add this action only if `@guardian.can_recover_post?`

More important is check when action is evaluated but displaying proper list of actions would be nice